File: figures/src/python/run_class_renderer.py
#!/usr/bin/env python3
"""Run a class-based Python renderer for a specific figure spec."""

from __future__ import annotations

import argparse
import importlib.util
from pathlib import Path
import sys


REPO_ROOT = Path(__file__).resolve().parents[3]
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))


def load_module(class_id: str):
    module_path = REPO_ROOT / f"figures/src/python/classes/{class_id}.py"
    if not module_path.exists():
        raise FileNotFoundError(f"Unknown Python class renderer: {class_id}")
    spec = importlib.util.spec_from_file_location(f"figure_class_{class_id}", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load renderer module for {class_id}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--class", dest="class_id", required=True)
    parser.add_argument("--spec", dest="spec_path", required=True)
    args = parser.parse_args()

    module = load_module(args.class_id)
    spec_path = (REPO_ROOT / args.spec_path).resolve()
    module.build_figure(spec_path)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: figures/src/r/run_class_renderer.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(jsonlite)
  library(systemfonts)
  library(yaml)
  library(ggplot2)
  library(patchwork)
  library(ragg)
  library(svglite)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 4) {
  stop("Usage: run_class_renderer.R --class <class_id> --spec <spec_path>")
}

class_flag <- args[[1]]
class_id <- args[[2]]
spec_flag <- args[[3]]
spec_path_arg <- args[[4]]
if (!identical(class_flag, "--class") || !identical(spec_flag, "--spec")) {
  stop("Usage: run_class_renderer.R --class <class_id> --spec <spec_path>")
}

script_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
if (length(script_arg) == 0) {
  stop("Unable to locate the current R script path")
}
script_path <- normalizePath(sub("^--file=", "", script_arg[[1]]), winslash = "/", mustWork = TRUE)
repo_root <- normalizePath(file.path(dirname(script_path), "..", "..", ".."), winslash = "/", mustWork = TRUE)
class_script <- file.path(repo_root, "figures", "src", "r", "classes", sprintf("%s.R", class_id))
if (!file.exists(class_script)) {
  stop(sprintf("Unknown R class renderer: %s", class_id))
}

options(manuscript.figure.repo_root = repo_root)
source(class_script)
build_figure(normalizePath(file.path(repo_root, spec_path_arg), winslash = "/", mustWork = TRUE))
